### DISCUSSION TITLE
Memory problems in Keycloak pods 25.0.1

### DISCUSSION THREAD

**Discussion by afmyr92** [keycloak/keycloak#33055] [Q&A]
Hello, 

I have a kubernetes cluster running two pods of keycloak with version 25.0.1. Both pods have as configured memory limit 6G and are started with parameters -XX:InitialRAMPercentage=50 and -XX:MaxRAMPercentage=70 in the JAVA_OPTS.
The service works correctly at startup, and gradually increases in memory. As the days go by and the memory of the pods reaches 4.2G (which coincides with approximately 70% of the limit in a week), they stop processing requests and the service drops.

I have checked the GC log (in my case G1GC) and it seems to be working correctly doing cleanups every <100ms, but I find it strange that the tendency of memory is always to grow, and the small memory spikes it releases come back up to higher values than before.

Regarding the cache, I use the default cache-ispn.xml file from the helm-chart. The volume of users keycloak is expected to manage is more than 50K users.

Some things I can think of are:
- Enable the persistent-user-sessions feature since it is not enabled by default in my version.
- Set limits on the infinispan cache since the default config comes with unlimited (most of the parameters are such that <distributed-cache name=“x” owners=“2”> <expiration lifespan=“-1”/>)

Would I be correct? Because after many changes I am not able to find where the problem is.

Thanks

**Comment by codespearhead** [comment]
Update to the latest Keycloak version and try again, because Infinispan was recently bumped to 15.0.7.Final, which was causing "a number of bugs affecting Keycloak users" as per https://github.com/keycloak/keycloak/issues/31963 .

**Comment by afmyr92** [comment]
I have updated several days ago to the latest version (25.0.6) and I am in the same scenario. Any other ideas?
Thanks!

**Reply by codespearhead** [reply]
Follow the steps in the description of this issue [1] to check if this applies to you. It might be a regression.

[1] https://github.com/keycloak/keycloak/issues/9877

**Reply by afmyr92** [reply]
I'm reading about issue #9877 and it says that ‘Offline users sessions are stored in infinispan cache and in DB’. In my case I was considering setting persistent-user-sessions feature, but it's not active yet, so I think all sessions are being stored in cache, so I don't really know how to reproduce it. In fact, I've made a query to the offline_client_session table of the postgres database I'm using and I get empty values: SELECT * FROM offline_client_session LIMIT 10; (0 rows)

To give more context to the problem the realm parameters configured through the interface are:
  _SSO Session Idle : 15 days
  SSO Session Max: 15 days
  Client Session Idle: 180 days
  Client Session Max: 180 days
  Offline Session Idle: 15 days
  Offline Session Max Limited: disabled
  Access Token Lifespan: 5 minutes_
  
 If I go inside the pod and check the /opt/keycloak/conf/cache-ispn.xml file this is the config regarding the distributed cache part:
`<distributed-cache name="sessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="authenticationSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="offlineSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="clientSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="offlineClientSessions" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>
        <distributed-cache name="loginFailures" owners="2">
            <expiration lifespan="-1"/>
        </distributed-cache>`
        
I'm wondering if these infinite values are the cause of my problem, but somewhere I read that the values in the cache-ispn.xml file are overwritten with the ones configured through the web interface I specified above.

**Reply by codespearhead** [reply]
Given the high values in your current setup, it's likely that Infinispan’s memory usage is spiraling out of control, which is ultimately causing your pods to crash.

Using an external Infinispan instance could serve as a temporary workaround [1], but this isn't a permanent fix unless the server running it has unlimited resources. If that were the case, you'd be able to stick with the embedded version of Infinispan alongside Keycloak on the same server.

Also, keep in mind that the Persistent User Sessions feature is expected to reach general availability in Keycloak 26 [2]. In the meantime, I’d recommend fine-tuning Infinispan [3], as outlined in this article [4], to better manage memory usage and avoid the issue you're likely facing.

[1] https://www.keycloak.org/high-availability/connect-keycloak-to-external-infinispan
[2] https://github.com/keycloak/keycloak/discussions/28271
[3] https://www.keycloak.org/server/caching (see section "Specify your own cache configuration file")
[4] https://dev.to/mohammedalics/optimizing-keycloak-caches-best-practices-for-embedded-and-external-infinispan-l9e

**Reply by afmyr92** [reply]
Ok, thanks for the information received. I'm thinking if it's worth waiting for version 26 in order to set up the Persistent Users Session feature directly as it will reach general availability and I think it can help me with my problem.

I have read in the documentation that: ‘Keycloak will default to a maximum of 10'000 entries for each of the caches sessions, clientSessions, offlineSessions, and offlineClientSessions if no other maximum size is configured in Keycloak's cache configuration XML file ..... With this feature enabled, Keycloak's memory usage might be reduced and the database usage may increase.’ [1].  

So, I'm left wondering if I should also adjust the Infinispam file (conf/cache-ispn.xml) in this case as it seems that the user persistence database manages the cached data up to 10000 entries. 

I also don't really understand what is the relation between the values declared in Infinispam (conf/cache-ispn.xml) and the ones configured through the keycloak administration panel ‘realm settings’ as both seem to be focused on the same purpose.

Could you shed some light on this? 
<img width="459" alt="image" src="https://github.com/user-attachments/assets/ff2621db-a039-486d-83a4-ff1597959c9a">

[1] https://www.keycloak.org/2024/06/persistent-user-sessions-in-preview

**Reply by codespearhead** [reply]
Could you close this discussion [1] now that the Persistent User Sessions feature [2] has reached GA in Keycloak 26?

[1] https://docs.github.com/en/discussions/managing-discussions-for-your-community/managing-discussions#closing-a-discussion
[2] https://github.com/keycloak/keycloak/blob/04e8508aaa48592b202a7d5a1c8737065b8435fa/docs/documentation/release_notes/topics/26_0_0.adoc?plain=1#L28

**Reply by afmyr92** [reply]
Yes, effectively setting up persistent sessions solved my problem.
